src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs

```diff
@@ -0,0 +1,408 @@
+        {
+            if (expression.Method.DeclaringType != typeof(Enumerable))
+                return null;
+
+            var type = expression.Arguments[0].Type;
+
+            if (!type.IsArray && type != typeof(string) && (!type.IsGenericType || type.GetGenericTypeDefinition() != typeof(List<>)))
+                return null;
+
+            if (!VersionAtLeast(7, 4))
+                return null;
+
+            switch (expression.Method.Name)
+            {
+            case nameof(Enumerable.Count) when type == typeof(string):
+                return
+                    Expression.Coalesce(
+                        new SqlFunctionExpression(
+                            "character_length",
+                            typeof(int?),
+                            new[] { expression.Arguments[0] }),
+                        Expression.Constant(0));
+
+            case nameof(Enumerable.Count):
+                return VersionAtLeast(8, 4)
+                    ? Expression.Coalesce(
+                        new SqlFunctionExpression(
+                            "array_length",
+                            typeof(int?),
+                            new[] { expression.Arguments[0], Expression.Constant(1) }),
+                        Expression.Constant(0))
+                    : null;
+
+            case nameof(Enumerable.ElementAt):
+                return MakeIndex(expression.Arguments[0], expression.Arguments[1]);
+
+            case nameof(Enumerable.Append):
+                return new CustomBinaryExpression(expression.Arguments[0], expression.Arguments[1], "||", expression.Arguments[0].Type);
+
+            case nameof(Enumerable.Prepend):
+                return new CustomBinaryExpression(expression.Arguments[1], expression.Arguments[0], "||", expression.Arguments[0].Type);
+
+            case nameof(Enumerable.SequenceEqual):
+                return Expression.MakeBinary(ExpressionType.Equal, expression.Arguments[0], expression.Arguments[1]);
+
+            default:
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Translates methods defined on <see cref="NpgsqlArrayExtensions"/>.
+        /// </summary>
+        /// <param name="expression">The expression to translate.</param>
+        /// <returns>
+        /// A translated expression or null if no translation is supported.
+        /// </returns>
+        /// <remarks>
+        /// This handler throws on unsupported versions since <see cref="NpgsqlArrayExtensions"/>
+        /// does not support client-side evaluation.
+        /// </remarks>
+        [CanBeNull]
+        Expression NpgsqlArrayExtensionsHandler([NotNull] MethodCallExpression expression)
+        {
+            if (expression.Method.DeclaringType != typeof(NpgsqlArrayExtensions))
+                return null;
+
+            switch (expression.Method.Name)
+            {
+            case nameof(NpgsqlArrayExtensions.Contains):
+                return VersionAtLeast(8, 2)
```

Comment:
(Sorry if we've already discussed this and I forgot)

I'm not sure we should throw here. The extension methods (e.g. `NpgsqlArrayExtensions.Contains`) already throw a ClientEvaluationNotSupportedException, so you can simply translate if the version is correct, and return null otherwise. This will trigger client evaluation which will throw.

In other words, if a database version doesn't support an operation, I think it's better to not translate and let client evaluation take its course.

This will also make code here more terse, as you can move the `VersionAtLeast()` check into the case and remove the conditional expression.

Reply:
Oh, that's a great point. I'll make the changes throughout.